[#250] Use value_counts() for PandasSheetFreqTable

azjps opened the pull request:
**do not merge yet**

Put in a bit of initial work into exploring using `value_counts()` for a Frequency Table for pandas. The surface area of the frequency table was considerably larger than I expected so I only implemented a small subset of features, but I'll note that a lot of the expected features weren't working prior either (for example, #266). Creating this PR for some initial feedback.

An initial hacky implementation for using value_counts()
to generate frequency table for pandas, which is vectorized
and hence significantly faster than the approach in
PivotTable. Added support for [Enter] > sourcerows.
Toggling/selection support is still broken here because
id(row) will not work with pandas.

Also tried to do some small cleanup with dtype
handling in PandasSheet.

This commit includes a lot of experimental and debugging
code which can be removed.

todo: support toggle in PandasSheet/PandasSheetFreqTable

azjps committed:
[#250] Use value_counts() for PandasSheetFreqTable

An initial hacky implementation for using value_counts()
to generate frequency table for pandas, which is vectorized
and hence significantly faster than the approach in
PivotTable. Added support for [Enter] > sourcerows.
Toggling/selection support is still broken here because
id(row) will not work with pandas.

Also tried to do some small cleanup with dtype
handling in PandasSheet.

This commit includes a lot of experimental and debugging
code which can be removed.

todo: support toggle in PandasSheet/PandasSheetFreqTable

anjakefala commented:
I really appreciate that you decided to open a WIP PR as the work started out, @azjps, because it helps us be more involved in the process along the way. 

Also, a general thank you for helping us improve the pandas loader. :blush: :blush: 

It is good timing because the Frequency Table / Pivot Table are also undergoing a design overview. It will be interesting to see how the pieces fit together. We will be able to take a more detailed look at the contents of your initial work, soon. =)

azjps commented:
Glad to assist, do let me know if there's anything I can clarify about this commit since it is pretty exploratory/messy. :)

saulpw commented on visidata/loaders/pandas_freqtbl.py:
```diff
@@ -0,0 +1,155 @@
+            self.addColumn(c)
+
+        # TODO: can make this part async w/ progress bar
+        for element in value_counts.index:
```
if `.index` has a len(), then you can use `Progress(value_counts.index)` and that will take care of it.  Otherwise `with Progress(total=n) as prog: ... prog.addProgress(1)`

saulpw commented on visidata/loaders/pandas_freqtbl.py:
```diff
@@ -0,0 +1,155 @@
+        df = self.source.rows
+
+        if len(self.groupByCols) > 1:
+            raise NotImplementedError("Frequency Table for > 1 columns currently not supported for PandasSheet")
```
I use `error()` or `fail()` for these (both of which raise an ExpectedException).  error() is red; fail() is yellow and less severe.

azjps replied:
👍 

I think it should be possible to do multiple `groupByCols` here by either using https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.crosstab.html and `stack()`, or by dropping down to `PivotTable`'s implementation. I'm not sure if there are gotchas to using `pd.crosstab().stack()` but I'll give it a shot.

saulpw commented:
Hey @azjps, I think you're on the right track here.  A faster and more functional PandasSheet would be appreciated by a bunch of people.

I did not even realize that selection doesn't work at all on pandas sheets. We'll probably have to overload the whole selection mechanism for PandasSheet.  What other selection mechanism would work for Pandas, do you think?

I also just cleaned up the selection API a bit, so at develop HEAD, the functions and properties from Sheet that would need to be overloaded are:

- `selectRow(sheet, row)`
- `unselectRow(sheet, row)`
- `unselectAll(sheet)`
- `isSelected(sheet, row)`: bool
- `selectedRows`: list of selected rows in sheet order
- `nSelected`: int

All other selection features should stem from those functions.

azjps commented:
Thanks for the comments, I'll look into cleaning those up when I get a chance.

Re: selection, unfortunately in the general case, I don't believe it is possible to uniquely identify a row of a pandas.DataFrame only given the row itself, as is done in `selectRow(sheet, row)`, namely if the index of the DataFrame contains non-unique entries. Non-unique index elements can be annoying to deal with in other places too though so it may very well be reasonable to require that the index is unique (or even stricter that it must be a default `RangeIndex`, e.g. after calling `reset_index()`). The other alternative I had in mind was to manipulate the row object to pass a bit more information such as the exact position of the row in the DataFrame. Either way I think the idea would be to rely on the row's position as opposed to the row object itself. A potential downside I see from this is selecting might interact poorly with editing (adding / removing rows), although I don't believe editing is supported currently for `PandasSheet`.

saulpw commented:
> I don't believe it is possible to uniquely identify a row of a pandas.DataFrame only given the row itself

Yeah, this sounds problematic.  VisiData uses id(row) as a really effective and general-purpose optimization strategy that has worked well, except for pandas.  I don't think we can get around the need for something like this (dealing with millions of rows, with potentially millions of rows selected, would otherwise require a substantial rearchitecture), but what if we at least made it easier to experiment?  Like calling a Sheet.id(row) method which by default returns the builtin id(), but in the case of a PandasSheet could return the index or some other annotation on the row?  This would allow the API to stay the same and require only a minimum change for other Sheets which can't use the builtin id.  I think your potential solutions would work within this framework, should we give it a try?

azjps commented:
Yeah that might work, I have a similar idea in mind that I'll give a shot implementing.

On an unrelated note, is there some way to view output logging or some other form of debugging? I've been debugging by "logging" the information I want in an exception message and using Ctrl+E to see it, but its a pretty inefficient workflow especially since I'm still pretty new to the code.

azjps committed:
[#250] Support for multi-column frequency table for PandasSheet

Use pandas.crosstab() to get multi-column frequencies.
This is not particularly space-efficient, and will also
not work for histogram-style numeric binning.

azjps committed:
[#266] Selection support for PandasSheet

Override the core selection API in PandasSheet,
keep track of selections using a bool dtype
pd.Series instead. Currently looking up rows
by index, which does have the requirement that
the DataFrame's index is unique.

The intended benefit is to be able to use vectorization
for multi-row selection, although this hasn't been
implemented yet.

saulpw commented:
Ah, that's a great question. Traditional printf-style debugging is more difficult in VisiData, it's true; I usually use `status()` (or `fail()` to do status-with-raise), which has a couple of nice features like msg dedup and a more efficient history with `Ctrl+P`.  I only use `Ctrl+E` for internal errors when I need to see the backtrace.

For development these days I get a lot more mileage out of `Ctrl+Y`, which lets you inspect the row object itself (or sheet object with `g Ctrl+Y`), and then you can drill down from there.  I'm thinking that might be what you're looking for.

Also, if I'm debugging something, I tend to create a .vd script and replay whatever commands I'm doing (whether the whole test or just the setup).  Then I have two windows, one for editor and one for visidata, and in the vd window I can rerun the test with `gq Up Enter` and it's a really quick turnaround time.

azjps commented on visidata/loaders/_pandas.py:
```diff
@@ -75,6 +79,34 @@ def sort(self):
+    @property
+    def selectedRows(self):
+        self._checkSelectedIndex()
+        return self.df.loc[self._selectedMask]
```
`selectedRows` is probably wrong here, the result doesn't have the right iterable output. Probably need to wrap in a DataFrameAdapter

saulpw replied:
ah yeah, or a different kind of adapter.  One thing to keep in mind, is that selectedRows should return a list of rows in the same order as the sheet itself.  This is for things like `dup-sheet` (`"`) to feel right. I think your approach already does this, but I wanted to make sure it's explicit.

azjps commented on visidata/loaders/pandas_freqtbl.py:
```diff
@@ -92,60 +92,76 @@ def unselectRow(self, row):
+            # have space on the order of product of number of unique elements for each
+            # column, even though its possible the combinations present are sparse
+            # and most combinations have zero count.
+            value_counts = pd.crosstab(this_column, [df.df[c.name] for c in self.groupByCols[1:]])
```
on second thought, it will likely make much more sense to use `pandas.pivot_table` for this. Possibly can even use to implement `PivotSheet`, albeit ambitious

saulpw commented:
Hey @azjps, how are you doing with this PR?  I was thinking, maybe it would be useful if we made a time to work on this together.  We could have a shared tmate session and either voice or text chat going at the same time.  With your pandas skills and my VisiData knowledge we might be able to get it working in fairly short order.  What do you think?

azjps committed:
[#266] Vectorized implementations for multi-row selection

Change "gu" to bind to unselect() to use public API.
Add methods for vectorized implementations for multi-row
selection and un-selection for PandasSheet. Fix
np.issubdtype() calls from #265.

azjps committed:
[#256] Fix bugs in previous commit

Fix to actually using new PandasSheet selection methods.

azjps commented:
Sounds like a good idea, I'll see when I can pull out some time to do such (will need to figure out if I can get tmate working on WSL first 😅 ). For this PR I still want to try changing crosstab() to pivot_table(), but other then that I'm pretty happy with it as far as my local testing goes. It would be good to get some unit tests added too which I'm not sure how to go about, as well as maybe trying a larger dataset for some quick benchmarks (which doesn't need to be committed, could just be some random data generated by a script).

saulpw commented on visidata/loaders/_pandas.py:
```diff
@@ -19,25 +21,43 @@ def __getattr__(self, k):
+                return int 
+            if np.issubdtype(dtype, np.floating):
+              return float
+            if np.issubdtype(dtype, np.datetime64):
```
does this work with datetime columns?  I'm not sure if I commented it out because it was the wrong mechanism (thanks for showing me the right idiom, btw), or because visidata.date and np.datetime64 aren't compatible.

azjps replied:
It does work with datetime columns, I work quite a bit with time series and have used them in visidata before 😄

saulpw commented on visidata/loaders/_pandas.py:
```diff
@@ -19,25 +21,43 @@ def __getattr__(self, k):
             self.df = readfunc(self.source.resolve(), **options('pandas_'+filetype+'_'))
 
-        self.columns = [ColumnItem(col, type=dtypeToType(self.df, col)) for col in self.df.columns]
+        # TODO: should we reset the index here and add it as a key column?
```
Ah, this is exactly what @danlat was suggesting in his PR.  I'm not so familiar with pandas DF and its index, but it sounds like it's something that pandas users want/need access to.  So let's do that (maybe merging them both in separately).

danlat replied:
It's similar to a database index column in that it's meant to be sorted and unique for quick access and querying, but pandas does not include the index in DataFrame.columns (from the list comprehension in this code)

saulpw commented on visidata/loaders/_pandas.py:
```diff
@@ -49,6 +69,57 @@ def sort(self):
+            # is no longer valid -- just delete it. Maybe too heavy but erring
+            # on side of correctness here.
+            vd.status('pd.DataFrame.index updated, clearing selected {} rows'
+                      .format(self._selectedMask.sum()))
```
Let's use `.nSelected` as you've redefined below. and a minor text change, to `{} selected rows`.

azjps replied:
It would lead to a cyclic reference here, since `nSelected` is calling `_checkSelectedIndex` itself, although the calling of `_checkSelectedIndex` is arguably overly defensive.

saulpw commented on visidata/loaders/_pandas.py:
```diff
@@ -61,3 +132,19 @@ def open_dta(p):
+undoSheetSelection = undoAttrCopy('[sheet]', '_selectedMask')
+
+# Override with vectorized implementations
+PandasSheet.addCommand('gt', 'stoggle-rows', 'toggleByIndex()', undo=undoSheetSelection),
```
I *think* these command redefinitions should be `None` for the keybinding, so that if someone binds `gt` globally to another longname it won't get rebound here on the PandasSheet.  But we'd have to make sure that `gt` etc still properly bind to the longname on the PandasSheet, of course (and I'll have to fix the keybinding code if they don't).

azjps replied:
Done, I haven't looked at the keybinding code but setting to `None` seems to work fine for what I've tried.

saulpw reviewed (commented):
I only had a few comments; I think if this is working well for you, let's clean it up (mostly removing commented out code and TODO comments) and merge it in and see how it works for other people.

Thanks so much for doing this, I don't have the experience with pandas to do it well, and I'm glad that the pandas loader is getting some love!

azjps committed:
[#256] Switch PandasFreqTable to use pd.pivot_table

1) Use pd.pivot_table which will be more memory-efficient
   than calling pd.crosstab().unstack(), which can be large
   if the cross-counts are sparse. The downside is that
   I haven't found any API for using pivot_table without
   adding a new column to the dataframe.
2) Change PandasFreqTable.select to directly interact with
   boolean selection index, which is significantly more
   performant than looping (when I tried it for a selection
   of ~100k rows, it took a several minutes to reach 1%;
   now it should be instantaneous).
3) Various small cleanups

azjps commented:
I've left some code comments in the hope that they will be instructive, since I wasn't 100% sure about some implementation choices I made. Besides that, I'm pretty satisfied with the state of these commits. Here's the snippet I used for testing / performance timings; it creates a 1 million row dataframe with various column dtypes. I wasn't sure where in the repository to include it, if at all:
```python
#!/usr/bin/env python3
"""
Run visidata PandasSheet on a dataframe with columns
of different datatypes and a length controlled by
the command-line option --repeat. The default size
of the loaded dataframe is 1 million rows.
"""

import argparse

import pandas as pd
import numpy as np

import visidata

visidata.options.disp_date_fmt = "%Y%m%d %H:%M:%S.%f"

if __name__ == "__main__":
    group_count = 10
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--repeat", default=100000, type=int,
        help="Number of times to repeat rows, total # of rows = {} * repeat. "
             "Default=100000 for a million rows total."
             .format(group_count)
    )
    args = parser.parse_args()

    repeat_count = args.repeat
    df = pd.DataFrame({
        "date1": pd.date_range(start="2020/01/01", freq="D", periods=group_count),
        "date2": pd.date_range(start="2020/01/01 02:15:30.012345", freq="D", periods=group_count),
        "int1": np.arange(group_count),
        "float1": np.arange(group_count) + 0.5,
        "cat1": [str(i) for i in np.arange(group_count)],
        "repeat1": np.repeat(1, group_count),
    })
    df["cat1"] = df["cat1"].astype("category")
    df["timedelta1"] = df["date1"].diff()
    # Repeat so that we are a significant amount of rows
    big_df = df.iloc[np.repeat(np.arange(0, group_count), repeat_count)]
    big_df = big_df.copy().reset_index(drop=True)
    # To make the frequencies a bit more interesting, shift the int2 & date2 columns 20% "upwards"
    big_df["int2"] = big_df["int1"].shift(-repeat_count // 5)
    big_df["date2"] = big_df["date2"].shift(-repeat_count // 5)
    big_df["int3"] = big_df["int1"].shift(repeat_count // 5)

    visidata.run(visidata.PandasSheet('pandas', source=big_df))
```

anjakefala commented:
Hi @azjps! 

# Your PR
Thank you so much for all of the work you have done on improving the pandas loader. As we were testing it, we discovered some performance issues which we are going to look into fixing after we merge. Once we make them, we will leave a comment outlining them, so you can see what we have done.

# Testing
Thanks so much for the testing script. 

We mostly have functional tests with `.vd` scripts. We are going to incorporate your python script into this setup, because testing on a pandas Dataframe would be super useful.
